### DISCUSSION TITLE
UE keep connect and disconnect

### DISCUSSION THREAD

**Discussion by initChat** [srsran/srsRAN_Project#590] [General Help] [answered]
I have followed this tutorial [srsRAN gNB with COTS UEs](https://docs.srsran.com/projects/project/en/latest/tutorials/source/cotsUE/source/index.html). **It seems that my UE can connect to network, but it keep being released all the time. So my UE cannot use internet.**

My system
- Ubuntu 22.04 to run open5gs by `./build/test/apps/5gc` which run config from [sample.yml.txt](https://github.com/srsran/srsRAN_Project/files/15161432/sample.yml.txt)
- B210 as RAN, which run under `performance mode` and run with `sudo ./gnb -c gnb_config.yaml` with config [gnb_config.txt](https://github.com/srsran/srsRAN_Project/files/15161433/gnb_config.txt)
- Quectel RM510 as UE
- the amf address is `127.0.0.2` 

The gNB can connect to the Core and receive IP address, as shown in [gnb_trace.txt](https://github.com/srsran/srsRAN_Project/files/15161457/gnb_trace.txt)
```
$ sudo ./gnb -c gnb_fllay.yml 

The PRACH detector will not meet the performance requirements with the configuration {Format B4, ZCZ 0, SCS 30kHz, Rx ports 1}.
Lower PHY in quad executor mode.

--== srsRAN gNB (commit 2f90c8b60) ==--

Connecting to AMF on 127.0.0.2:38412
Available radio types: uhd.
[INFO] [UHD] linux; GNU C++ version 10.5.0; Boost_107400; UHD_4.0.0.HEAD-0-g90ce6062
[INFO] [LOGGING] Fastpath logging disabled at runtime.
Making USRP object with args 'type=b200,num_recv_frames=64,num_send_frames=64'
[INFO] [B200] Detected Device: B210
[INFO] [B200] Operating over USB 3.
[INFO] [B200] Initialize CODEC control...
[INFO] [B200] Initialize Radio control...
[INFO] [B200] Performing register loopback test... 
[INFO] [B200] Register loopback test passed
[INFO] [B200] Performing register loopback test... 
[INFO] [B200] Register loopback test passed
[INFO] [B200] Setting master clock rate selection to 'automatic'.
[INFO] [B200] Asking for clock rate 16.000000 MHz... 
[INFO] [B200] Actually got clock rate 16.000000 MHz.
[INFO] [MULTI_USRP] Setting master clock rate selection to 'manual'.
[INFO] [B200] Asking for clock rate 23.040000 MHz... 
[INFO] [B200] Actually got clock rate 23.040000 MHz.
Cell pci=1, bw=20 MHz, dl_arfcn=627340 (n78), dl_freq=3410.1 MHz, dl_ssb_arfcn=626976, ul_freq=3410.1 MHz

==== gNodeB started ===
```

However, when I try to connect UE to the system, it gets registered and released repeatedly, as shown on log of open5gs [5gc.log](https://github.com/srsran/srsRAN_Project/files/15161448/5gc.log)
```
04/24 11:34:06.125: [amf] INFO: InitialUEMessage (../src/amf/ngap-handler.c:401)
04/24 11:34:06.125: [amf] INFO: [Added] Number of gNB-UEs is now 1 (../src/amf/context.c:2656)
04/24 11:34:06.125: [amf] INFO:     RAN_UE_NGAP_ID[0] AMF_UE_NGAP_ID[1] TAC[7] CellID[0x66c000] (../src/amf/ngap-handler.c:562)
04/24 11:34:06.125: [amf] INFO: Unknown UE by 5G-S_TMSI[AMF_ID:0x20040,M_TMSI:0xc00005e2] (../src/amf/context.c:1858)
04/24 11:34:06.125: [amf] INFO: [Added] Number of AMF-UEs is now 1 (../src/amf/context.c:1621)
04/24 11:34:06.125: [gmm] INFO: Registration request (../src/amf/gmm-sm.c:1224)
04/24 11:34:06.125: [gmm] INFO: [Unknown ID]    5G-S_GUTI[AMF_ID:0x20040,M_TMSI:0xc00005e2] (../src/amf/gmm-handler.c:185)
04/24 11:34:06.145: [gmm] INFO: Identity response (../src/amf/gmm-sm.c:1492)
04/24 11:34:06.145: [gmm] INFO: [suci-0-001-01-0-0-0-0000062714]    SUCI (../src/amf/gmm-handler.c:944)
04/24 11:34:06.432: [gmm] INFO: [imsi-001010000062714] Registration complete (../src/amf/gmm-sm.c:2323)
04/24 11:34:06.432: [amf] INFO: [imsi-001010000062714] Configuration update command (../src/amf/nas-path.c:591)
04/24 11:34:06.432: [gmm] INFO:     UTC [2024-04-24T04:34:06] Timezone[0]/DST[0] (../src/amf/gmm-build.c:558)
04/24 11:34:06.432: [gmm] INFO:     LOCAL [2024-04-24T11:34:06] Timezone[25200]/DST[0] (../src/amf/gmm-build.c:563)
04/24 11:34:09.262: [amf] INFO: UE Context Release [Action:2] (../src/amf/ngap-handler.c:1696)
04/24 11:34:09.262: [amf] INFO:     RAN_UE_NGAP_ID[0] AMF_UE_NGAP_ID[1] (../src/amf/ngap-handler.c:1697)
04/24 11:34:09.262: [amf] INFO:     SUCI[suci-0-001-01-0-0-0-0000062714] (../src/amf/ngap-handler.c:1701)
04/24 11:34:09.262: [amf] INFO: [Removed] Number of gNB-UEs is now 0 (../src/amf/context.c:2663)
```

With unstable connection, the UE cannot access internet although it can connect to the network as shown on quectel minicom [quectel_minicom.txt](https://github.com/srsran/srsRAN_Project/files/15161398/quectel_minicom.txt)
```
at+cops?                                                                     
+COPS: 0,0,"Open5GS Magic",11    
```

Also the wireshark of gnb only return `mac` but `ngap` files return blank [gNB_wireshark.zip](https://github.com/srsran/srsRAN_Project/files/15161411/gNB_wireshark.zip)

I have tried
1. make sure to Enable IPv4/IPv6 Forwarding and add NAT rule
```
$ sudo sysctl -w net.ipv4.ip_forward=1
$ sudo sysctl -w net.ipv6.conf.all.forwarding=1

### Add NAT Rule
$ sudo iptables -t nat -A POSTROUTING -s 10.45.0.0/16 ! -o ogstun -j MASQUERADE
$ sudo ip6tables -t nat -A POSTROUTING -s 2001:db8:cafe::/48 ! -o ogstun -j MASQUERADE
```
2. config sim as `90170` with `pySim-prog.py`. I used to use `00101` which is not successful as well.
3. disable service 124 and 125 with `pySim-shell.py` with script
```
verify_adm XXXXX
# Key Provisioning
select MF
select ADF.USIM 
select DF.5GS 
select EF.SUCI_Calc_Info 
update_binary_decoded '{ "prot_scheme_id_list": [ {"priority": 0, "identifier": 2, "key_index": 1}, {"priority": 1, "identifier": 1, "key_index": 2}, {"priority": 2, "identifier": 0, "key_index": 0}], "hnet_pubkey_list": [ {"hnet_pubkey_identifier": 27, "hnet_pubkey": "0272DA71976234CE833A6907425867B82E074D44EF907DFB4B3E21C1C2256EBCD1"}, {"hnet_pubkey_identifier": 30, "hnet_pubkey": "5A8D38864820197C3394B92613B20B91633CBD897119273BF8E4A6F4EEC0A650"}]}'

# Routing Indicator
select MF
select ADF.USIM 
select DF.5GS 
select EF.Routing_Indicator
update_binary 0071ffff

# Service Table
select MF
select ADF.USIM 
select EF.UST 
ust_service_deactivate 124
ust_service_deactivate 125
```
4. change ue from Huawei CPE to Quectel
5.  with Quectel, I follow this [paper](https://www.eurecom.fr/fr/publication/7497) and set it to use `IPv4`, make it `roaming`, make it connect to APN `internet` which i specify on `localhost:9999`
```
AT+QMBNCFG=“select”,“ROW Commercial”
2) AT+QNWPREFCFG=“nr5g band”
3) AT+QNWPREFCFG=“mode pref”
4) AT+QNWPREFCFG=“mode pref”, nr5g
5) AT+QNWPREFCFG=“nr5g disable mode”,0
6) AT+CGDCONT=1,“IP”,“internet”,“0.0.0.0”,0,0
AT+QNWPREFCFG="roaming_pref",255
```

I have attached
**config files** 
[gnb_config.txt](https://github.com/srsran/srsRAN_Project/files/15161506/gnb_config.txt)
[sample.yml.txt](https://github.com/srsran/srsRAN_Project/files/15161508/sample.yml.txt)
[quectel_minicom.txt](https://github.com/srsran/srsRAN_Project/files/15161515/quectel_minicom.txt)

**log gnb**
[gnb_trace.txt](http
[gNB_wireshark.zip](https://github.com/srsran/srsRAN_Project/files/15161511/gNB_wireshark.zip)
s://github.com/srsran/srsRAN_Project/files/15161510/gnb_trace.txt)
**log open5gs**
[5gc.log](https://github.com/srsran/srsRAN_Project/files/15161513/5gc.log)

**Comment by pgawlowicz** [comment]
Could you provide gnb logs?

**Reply by initChat** [reply]
my gnb log turns out blank [gnb.log](https://github.com/srsran/srsRAN_Project/files/15161614/gnb.log)

**Reply by pgawlowicz** [reply]
Change log level to debug.

**Reply by initChat** [reply]
[debug_gnb.zip](https://github.com/srsran/srsRAN_Project/files/15161765/debug_gnb.zip)

**Comment by initChat** [comment]
@pgawlowicz I change to another UE. The UE can connect to the network consistently without being released, with a stable RNTI of `4601`.**However, the UE still cannot access the internet.**

I used spectrum analyzer to capture at center freq 3410 MHz. It shows spectrums. It peaks at 3416 MHz with amplitude -17.67 dBm.
![020524_SDR210](https://github.com/srsran/srsRAN_Project/assets/40858099/1f7b277a-0d96-4ee6-98cd-8707eefadc56)

**Could you please suggest what should I try more? Thank you very much.**

today's log (the code is done with the same config of this discussion)
I am sorry that the gNB log is quite hugh. So i post it on google drive. So please download it.
- gnb https://drive.google.com/file/d/17YaRBFZdKfkGMpkAmmgp_HU_ZqzT2Wy7/view?usp=drive_link
- gnb_wireshark (only MAC is presented, NGAP does not contains anything) [gnb_wireshark.zip](https://github.com/srsran/srsRAN_Project/files/15186305/gnb_wireshark.zip)
- open5gs [5gc.log](https://github.com/srsran/srsRAN_Project/files/15186311/5gc.log)

**Reply by pgawlowicz** [reply]
Can the UE ping core network?

Did you enable IP forwarding and add the NAT rule? see here: https://open5gs.org/open5gs/docs/tutorial/01-your-first-lte/

**Reply by initChat** [reply, marked as answer]
Thank you for reminding the NAT rule again. It turns out that my UE (ZLTX21 WIFI 5 CPE) need to be toggled on the NAT rule again before the UE can access the internet. I use iphone 13 connected to WiFi from this CPE and can test speedtest, IPerf properly. 
![ue_nat](https://github.com/srsran/srsRAN_Project/assets/40858099/6cadc8d5-c10f-4ece-910e-2ded48904793)

Thank you very much for your help! @pgawlowicz  🙏🙏🙏

**Comment by pengnist** [comment]
I also experience RRC release with srsue. But it only happens a few mins after connection is established and no application is launched, such as iperf or ping. Is this expected?

**Reply by HaoxinSEU** [reply]
Maybe you can set the `inactivity_timer` under `cu_cp` to 7200, so that the UE doesn't disconnect in a short time.

**Reply by pengnist** [reply]
@HaoxinSEU Problem solved. Thanks!

**Reply by initChat** [reply]
@HaoxinSEU This one is helping a lot! It gives me time to config my UE properly.
